### PR TITLE
[MLIR] Add f8E8M0FNU type

### PR DESCRIPTION
**Pull request by sergey-kozub**
This PR adds `f8E8M0FNU` type to MLIR.

`f8E8M0FNU` type is proposed in [OpenCompute MX Specification](https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf). It defines a 8-bit floating point number with bit layout S0E8M0. Unlike IEEE-754 types, there are no infinity, denormals, zeros or negative values.

```c
f8E8M0FNU
- Exponent bias: 127
- Maximum stored exponent value: 254 (binary 1111'1110)
- Maximum unbiased exponent value: 254 - 127 = 127
- Minimum stored exponent value: 0 (binary 0000'0000)
- Minimum unbiased exponent value: 0 − 127 = -127
- Doesn't have zero
- Doesn't have infinity
- NaN is encoded as binary 1111'1111

Additional details:
- Zeros cannot be represented
- Negative values cannot be represented
- Mantissa is always 1
```

Related PRs:
- [PR-107127](https://github.com/llvm/llvm-project/pull/107127) [APFloat] Add APFloat support for E8M0 type
- [PR-105573](https://github.com/llvm/llvm-project/pull/105573) [MLIR] Add f6E3M2FN type - was used as a template for this PR
- [PR-107999](https://github.com/llvm/llvm-project/pull/107999) [MLIR] Add f6E2M3FN type
- [PR-108877](https://github.com/llvm/llvm-project/pull/108877) [MLIR] Add f4E2M1FN type

### CODE REVIEW

**Comment by durga4github** [comment]
Looks good to me. Follows the same template from the earlier PRs.

In the commit message:
Unlike IEEE-754 types, there are no infinity, zeros or negative values

Can we also add "no denorms" too?

**Review by durga4github — APPROVED** [review]

**Review by dcaballe — APPROVED** [review]

**Comment by stellaraccident** [comment]
Fly on the wall, but at what point, upon removing all features that traditionally make something a "floating point number" (mantissa, zero, denorms, infinities) does something no longer make any sense at all being part of a floating point hierarchy. It's just a bit-vector with a special error value.

I'm not blocking this in any way or even asking seriously. Just kind of balking at the cargo cult mentality that is going into labeling bundling these things together like this.

**Comment by River707** [comment]
> Fly on the wall, but at what point, upon removing all features that traditionally make something a "floating point number" (mantissa, zero, denorms, infinities) does something no longer make any sense at all being part of a floating point hierarchy. It's just a bit-vector with a special error value.
> 
> I'm not blocking this in any way or even asking seriously. Just kind of balking at the cargo cult mentality that is going into bundling these things together like this.

I had the same gut reaction... We are now up to ~18 floating point types... That kind of points to a serious issue with the way things are scaling here, and I think we should really rethink what's being done (especially given that each one of these PRs are identical, add big chunks of code in the core library, etc).

**Comment by sergey-kozub** [comment]
> Fly on the wall, but at what point, upon removing all features that traditionally make something a "floating point number" (mantissa, zero, denorms, infinities) does something no longer make any sense at all being part of a floating point hierarchy. It's just a bit-vector with a special error value.

I also wondered what makes it a floating point number.
From my perspective, both int and floating point types represent values in a numeric range. The difference is that for int types, the distance between adjacent numbers is constant, and for floating point types it's variable.

For E8M0, mantissa is there but is implicit (has 1 bit which has value of one) - other FP types also have an implicit bit of data.
Unsigned ints also can't represent negative numbers (same with E8M0). Not having infinities is also common, e.g. for other FP8 types like E4M3FN and E5M2FNUZ.

The E8M0 is intended to be used as a scaling factor in block scaled formats like MXFP8, which is exactly why it doesn't have negatives, infinities or zeros - none of these makes sense for a scaling factor.

**Comment by stellaraccident** [comment]
> > Fly on the wall, but at what point, upon removing all features that traditionally make something a "floating point number" (mantissa, zero, denorms, infinities) does something no longer make any sense at all being part of a floating point hierarchy. It's just a bit-vector with a special error value.
> > I'm not blocking this in any way or even asking seriously. Just kind of balking at the cargo cult mentality that is going into bundling these things together like this.
> 
> I had the same gut reaction... We are now up to ~18 floating point types... That kind of points to a serious issue with the way things are scaling here, and I think we should really rethink what's being done (especially given that each one of these PRs are identical, add big chunks of code in the core library, etc).

Ok, I didn't want to unilaterally hit pause, but it seems like we've got a number of people with the same analysis. Should we at least discuss this a bit more? Or proceed with this patch? I agree that the path we're on is not very sustainable.

**Comment by sergey-kozub** [comment]
> especially given that each one of these PRs are identical

The amount of boilerplate code is annoying, I believe this could (and should) be generalized.
One could come up with a few dozen more FP8 (or smaller dtypes).

**Comment by stellaraccident** [comment]
> > especially given that each one of these PRs are identical
> 
> The amount of boilerplate code is annoying, I believe this could (and should) be generalized. One could come up with a few dozen more FP8 (or smaller dtypes).

Yeah, it was never meant to scale beyond the primary fp8 types. Needs a rethink... if not in this case, certainly soon.

**Comment by sergey-kozub** [comment]
> One could come up with a few dozen more FP8 (or smaller dtypes).

For example, https://arxiv.org/html/2405.13938v1 mentions more esoterics like `e0m3` and `e1m3`.

**Comment by stellaraccident** [comment]
> > One could come up with a few dozen more FP8 (or smaller dtypes).
> 
> For example, https://arxiv.org/html/2405.13938v1 mentions more esoterics like `e0m3` and `e1m3`.

Just judging by the mood and temperament of the industry, we'll end up with just about every combination before too long. Might as well try to structure the code for that eventuality vs being the victim of it.

**Comment by dcaballe** [comment]
Yeah, we’re definitely dealing with a combinatorial explosion. For the compiler internals, have we thought about using a single FP type that models different E/M sizes and supports the required feature flags for the specific semantics of these formats? I wonder, though, if we need verification at some level (bindings?) to ensure compliance with the micro-scaling standard... Or do we want to leave the door open to arbitrary format? I'm not sure I fully understand the implications of the latter.

> but at what point, upon removing all features that traditionally make something a "floating point number" 

I would say they are still floating-point in spirit :). I’m not sure we could treat them differently. There are specific FP semantics that still apply to these formats, such as NaN propagation (e.g., minnum/maxnum/minimum/maximum), exceptions, overflow...

**Comment by kuhar** [comment]
> I would say they are still floating-point in spirit :). I’m not sure we could treat them differently. There are specific FP semantics that still apply to these formats, such as NaN propagation (e.g., minnum/maxnum/minimum/maximum), exceptions, overflow...

Is there any hardware that supports arithmetic over these types? At least on the *CDNA* side, fp8 support is limited to conversion to/from other 'usual' fp types and dedicated matmul intrinsics. If we had IR that did any arithmetic on fp types, we wouldn't be able to compile it without software emulation anyway. Or in other words, I wonder if some level of *general* hw support (not just conversions + matmul) would be a good criteria for inclusion in core MLIR types.

**Comment by stellaraccident** [comment]
> Yeah, we’re definitely dealing with a combinatorial explosion. For the compiler internals, have we thought about using a single FP type that models different E/M sizes and supports the required feature flags for the specific semantics of these formats? I wonder, though, if we need verification at some level (bindings?) to ensure compliance with the micro-scaling standard... Or do we want to leave the door open to arbitrary format? I'm not sure I fully understand the implications of the latter.

I can't defend my opinion rigorously, but I think that many of these "software only" types are leaking in from a well intentioned reading of the MX spec and an attempt to mirror that in MLIR. But the MX spec has very little to say about how the machines implement such things -- and the early state of implementations means that few of us fine purveyors of such machines are saying a lot of concrete details about the underlying implementations yet. My observation is that this has made it hard to plan the right level of support in MLIR, as the norm in this ecosystem is to work bottom up. But that is hard to do when the bottom is not defined in a way that everyone can see. And that is probably causing (again well-intentioned) premature generalization.

As Jakub says, many of these things are going to get defined in terms of conversion intrinsics (or emulation/helper functions) and a very small number of ways that they can actually interact with the hardware. I'd have to go dig through the original discussion on FP8, but a key point of that was that those types had a concrete realization in hardware and that the compiler was better off fully modeling such a thing (vs type punning, etc). I'm not sure that the same line of reasoning necessarily holds for all of these software/emulation variants.

But like I said, I'm not sure I can defend that position rigorously. In the absence of that, I think my core concern is that we may want a way in the code to better model software/emulated FP types.

**Comment by sergey-kozub** [comment]
> I wonder if some level of _general_ hw support

The MX types (E2M1, E2M3, E3M2, E8M0) have hw support from NVidia.
The other types introduced recently (E3M4 https://github.com/llvm/llvm-project/pull/101230, E4M3 with inf https://github.com/llvm/llvm-project/pull/97118) have hw support from Amazon.

**Comment by stellaraccident** [comment]
Everyone has some degree of intrinsic support for the MX types, just like most things. But the nature of that support is pretty different from typical FP types.

In any case, I'm not against adding these types or even landing this PR. But I think the approach needs adjustment before this gets further out of hand.

**Comment by stellaraccident** [comment]
Started a thread: https://discourse.llvm.org/t/rethink-on-approach-to-low-precision-fp-types/82361

File: mlir/test/IR/attribute.mlir
Comment (jpienaar) [inline mlir/test/IR/attribute.mlir:1]: Note: this doesn't seem correctly roundtrip through bytecode. If checking IR/attribute.mlir

"test.float\_attrs"() \<{float\_attr = 2.000000e+00 : f8E8M0FNU}> : () -> () loc(#loc12)

becomes

"test.float\_attrs"() \<{float\_attr = 4.000000e+00 : f8E8M0FNU}> : () -> () loc(#loc12)

Can test with `MLIR_OPT_CHECK_IR_ROUNDTRIP=1` in testing environment.

Comment (sergey-kozub) [reply mlir/test/IR/attribute.mlir:1]: Fix in https://github.com/llvm/llvm-project/pull/113298

**Pull request merged**